feat(HMR)!: integrate Vite 6 HMR -> only run code generation pipeline if modfied file has graphql dependency, reduce HMR event count

rgon opened the pull request:
Fixes #1454

This PR refactors the dev server file watcher so it runs from standard Vite HMR events. It then only runs the houdini code generation pipeline if the file imports from the `$houdini` (or `config.runtimeDir`) path. It finally only asks for module replacement if there are changes in the generated code.

It reduces the amount of HMR events on any change in a project which imports Houdini from ~100 to 1, being less of a toll on the client browser.

It speeds up HMR events on the project: only slowing down those that really have graphql code niside.

This will set an easier base for fully atomic rebuilds.
 
---

#### Current build pipeline on dev server
On every change, the `houdini/src/codegen/index.ts compile` method is called from `houdini/src/vite/index.ts watchAndRun`.
https://github.com/HoudiniGraphql/houdini/blob/aff81e3698099270c2cccd315b2679f3061564ae/packages/houdini/src/vite/index.ts#L29C1-L33C40

```js
async run(server, absolutePath:string|null)
```

#### Proposed alternative
This `watchAndRun` method appears not to be necessary. Vite provides the [`hotUpdate` handler](https://vite.dev/guide/api-environment-plugins.html#the-hotupdate-hook) for create/update/delete, with a list of modules (useful later!) which are used to detect whether the file change should trigger a houdini rebuild or not.

https://github.com/HoudiniGraphql/houdini/blob/aff81e3698099270c2cccd315b2679f3061564ae/packages/houdini/src/vite/houdini.ts#L25

We just need to move the contents of the `watchFile` and `run` methods to this new handler.

#### Potential issues
Since HMR detection is handled separately from actual builds (which are on `buildStart`), this change probably does no damage. However, if there were a way to test HMR builds programatically, it would be smart. This is my first dive into both vite and houdini, so I need a hand on this!


## Task list/PR Status
+ [x] Find which function triggers the rebuild, and who rebuilds any file.
+ [x] Implement `hotUpdate` alternative ways of listening to file changes
+ [x] Cleanup: if working, deprecate watchAndRun and their dependencies
+ [x] Trigger HMR replacement only if something has changed
+ [x] Invalidate & successfully trigger update on houdini files
+ [x] detect if the file has a houdini dependency before calling build

rgon commented:
#### Path towards fine-grained builds
TODO
By using the `hotUpdate` handler, we get access to the vite dependency graph.

>  When a file is modified, the module graph of each environment will be used to discover the affected modules and perform HMR for each environment independently.

If detected, the list could probably get passed onto Houdini's `compile` method, which gets the document list from `collectDocuments`.

rgon commented:
Aparently, any changes on the `hotUpdate` hook are ignored by HMR file change detection (adding a log statement in the new `hotUpdate` hook after checks): 

Editing a file with graphql (or not):
```
🔄 Hot reloading modules:...
  - /app/apiviews/src/lib/user/SmartAddressInput.svelte
[houdini] Dependencies: []
1:37:31 PM [vite] (client) hmr update /@fs/app/apiviews/src/lib/user/SmartAddressInput.svelte, /src/app.postcss
1:37:31 PM [vite] (ssr) page reload /app/apiviews/src/lib/user/SmartAddressInput.svelte
```

Editing the file and changing the query within:
```
🎩 ✏️  GetUserAddresses
🔄 Hot reloading modules:...
  - /app/apiviews/src/lib/user/SmartAddressInput.svelte
[houdini] Dependencies: []
1:41:06 PM [vite] (client) hmr update /@fs/app/apiviews/src/lib/user/SmartAddressInput.svelte, /src/app.postcss
1:41:06 PM [vite] (ssr) page reload /app/apiviews/src/lib/user/SmartAddressInput.svelte
``` 

Well, the first issue of 100 HMR events per change described in https://github.com/HoudiniGraphql/houdini/issues/1454 appears to be fixed! Also, Houdini's own way to detect for changes is performed, which is excellent.

Now, the issue is that there's no reload for codegenerated files, so changing a query doesn't actually send the change (not even in explicit definitions `const query = graphql()`. We need to add a way to fetch the tainted files from `houdini/src/codegen/index.ts compile()` (when  `🎩 ✏️ ` is logged)

rgon commented:
I have succesfully implemented the reloading of houdini built files (by walking over the houdini directory, probably not ideal). I have tried to make this as decoupled as possible from the actual codegen pipeline, since it will be rewritten. 

Once atomic rebuilds are implemented, the walked list can be swapped with the tainted files.

As of now, this PR achieves:
+ No spurious HMR events on the codegenerated files: one change = 1 event. Less log spamming.
+ Only trigger the reload of the houdini modules in the browser if they have changed. Performance improvement (especially when running the Web Inspector).
+ Use Vite's own file change detection, instead of a separate one
+ Cleaner path towards actual atomic rebuilds

It does not:
+ Rebuild the entire houdini codegen folder when files that actually contain graphql are modified. However, this method should probably be implemented in the context of the new build pipeline.

I kindly ask for some testing/review from anyone in the houdini team before continuing, to see if this is this path aligns with future development!

rgon commented:
There's the issue that the code generation pipeline is run twice, because there's a hot reload event per each environment (client and ssr):
```
Reloading /houdini-classundefined-error-repro/src/routes/DummyGraphQLFetcher.svelte... on client env
🎩 🔄 bundle HMR rebuild...
Reloading /houdini-classundefined-error-repro/src/routes/DummyGraphQLFetcher.svelte... on ssr env
🎩 🔄 bundle HMR rebuild...
```
To fix this, we should check if the event's timestamp is the same as the previous one (from the last hot update event). Since the source event is the same, the timestamp matches. Fixed in the following commit.

AlecAivazis commented:
hey @rgon! I finally found some time to pull down your fork and ran the e2e tests. Unfortunately it seems that the dependency logic isn't working when I change a `+page.gql` file. Determining that dependency for `+layout.gql` files will be slightly tricky without some kind of coarse heuristic like every file underneath them is considered changed

rgon commented:
That makes sense, by the time the hot update event gets triggered, it most likely is treated just like a binary, with no module dependencies.

The easy way out is to bypass the $houdini dependency check if the file is *.gql and proceed with generation, since by definition they all require $houdini, but that means having no info for the incremental compilation pipeline. I'll have a look.

AlecAivazis commented:
could your branch maybe be out of date from main? there was a recent fix pushed up for this so maybe you can pull that in

rgon commented:
Thanks Alec,that was indeed the issue. It, however, did uncover that the hook is being called on `.houdini` and `.svelte-kit` changes, which it shouldn't (huge performance sink). Weirdly enough, this is only happening on the e2e project and not the rest I tried. One moment till I debug

SeppahBaws commented:
Huh strange, from what I can see the `.graphqlrc.yml` file is only ever used to set up IDE integration, so that the LSP knows about our generated documents...

rgon commented:
@SeppahBaws whoops big slip-up on my part then, shouldn't blindly trust search. The problematic code is:
```ts
export async function shouldReactToFileChange(
	filepath: string,
	opts: PluginConfig,
	watchSchemaListref: WatchSchemaType
): Promise<boolean> {
	const config = await getConfig(opts)

	// we need to watch some specific files
	if (config.localSchema) {
		const toWatch = watchSchemaListref.list
		if (toWatch.includes(filepath)) {
			// if it's a schema change, let's reload the config
			await getConfig({ ...opts, forceReload: true })
			return true
		}
	} else {
		const schemaPath = path.join(path.dirname(config.filepath), config.schemaPath!)
		if (minimatch(filepath, schemaPath)) {
			// if it's a schema change, let's reload the config
			await getConfig({ ...opts, forceReload: true })
			return true
		}
	}

	return config.includeFile(filepath, { root: process.cwd() })
}
```
The `config.localSchema` branch is only triggered on the e2e project, not a newly created one, nor any other installations I have tested it on. Thus including events on the generated folder.

I'll review it, even if the symptoms are fixed, just to be sure there isn't some other code I can reuse. Can't promise I'll have it today, though.

rgon commented:
It was a direct bug of the `isGraphql` check I introduced: if a file is graphql, the dependency check should be ignored (graphQL files always depend on Houdini), but the `shouldReactToFileChange` (houdini file watching rules) should not be ignored. The `isInGenerated` check was superfluous and buggy (`.startswith()` is not a good directory test for many reasons).

I believe it was only caused in the `e2e/kit` project probably because of a different folder structure (links+relative dirs) than the dockerized test I was using before, getting triggered on `.houdini/graphql/schema.graphql` and `.houdini/graphql/documents.gql`.

Fixed!
```
[web] 🎩 🔄 bundle HMR rebuild...
[web] 🎩 ✏️  ListAll
[web] 🎩 ⬆️ bundle changed, triggering HMR update
[web] 5:50:01 PM [vite] (client) page reload src/routes/lists/all/+page.gql
```

rgon commented:
This PR is now ready for review & up to date with main, so give me a head's up if there's anything off. I am currently using it in development in one of my projects.

AlecAivazis commented on packages/houdini/src/vite/hmr.ts:
```diff
@@ -0,0 +1,60 @@
+
+	// we need to watch some specific files
+	if (config.localSchema) {
+		const toWatch = watchSchemaListref.list
```
where does this value get set? If there is a local schema then the value should be `src/api/+schema`

rgon replied:
At a first glance, I'm not immediately sure. This is one of the segments that's moved from the original implementation (was on [`watchFile`](https://www.kitql.dev/docs/tools/03_vite-plugin-watch-and-run), now is in a separate guard function `shouldReactToFileChange`):

https://github.com/HoudiniGraphql/houdini/blob/64fe7964266773f0b3eda4ee779481c811b6bfae/packages/houdini/src/vite/index.ts#L33-L58

rgon replied:
It was added in #1203. The value is set here:

https://github.com/HoudiniGraphql/houdini/blob/64fe7964266773f0b3eda4ee779481c811b6bfae/packages/houdini/src/lib/config.ts#L1134-L1145

Still figuring out why this check is performed

rgon replied:
As far as I understand, this check seems superfluous, yet I don't fully understand why it was added on the first place.

The value of `watchSchemaListRef` is set on `moduleParsed`.

https://github.com/HoudiniGraphql/houdini/blob/64fe7964266773f0b3eda4ee779481c811b6bfae/packages/houdini/src/vite/schema.ts#L89-L112

But the Vite docs explain
> [moduleParsed](https://rollupjs.org/plugin-development/#moduleparsed) hook is not called during dev, because Vite avoids full AST parses for better performance.
> [source](https://v5.vite.dev/guide/api-plugin.html#universal-hooks)

Since these schema watchers are only used on the dev server, we don't need this... right?

----

Since the value that gets added to `watchSchemaListRef` in `houdini/src/vite/schema.ts` is basically:
```ts
if (module.id.endsWith('+schema.ts')) {
  depOfSchema.push(module.id, ...module.importedIdResolutions.map((c) => c.id))
}
```

Removing all mentions to `watchSchemaListRef` and swapping L25 from
```ts
if (config.localSchema) { 
```
to
```ts
if (minimatch(filepath, config.localSchemaPath)) {
```
seems sensible.

@AlecAivazis are there any existing e2e test for localSchema or some other project I can test this on? In projects with remote schemas, this change doesn't matter, including `e2e/kit`, where watchSchemaListref is always blank:
```
watchSchemaListref { list: [] }
```

rgon replied:
No worries, I found out and tested it in the meantime. This whole approach doesn't work with Vite v4, which the e2e tests were using. However, Vite v4 is not supported by Houdini, so a plain update to the e2e tests is enough (no major version change required). Created a PR for it #1466

rgon commented:
Current issues:
+ [ ] Approach requires Vite 6. Houdini sets peerDependency as vite 5||6. Supporting Vite 5 would require switching the `hotUpdate` plugin hook to the (non-environment aware `handleHotUpdate` hook.
+ [ ] Not working on react e2e test (it uses the unsupported Vite 4 dependency)
+ [ ] Resolve watchSchemaListRef issue

AlecAivazis commented:
@rgon i'd like to try to merge this into the houdini-2.0 branch and deploy a prelease to see if it fixes some HMR issues that people are having with 6.0. I would have expected you to run into the issues while debugging this branch and it doesn't sound like you have.

is this ready to go if we are willing to hard-commit on vite6?

rgon commented:
@AlecAivazis yes, it's ready to go! I have been using this patched version on a real project (in dev) for a week or so with no issues (vite 6.2.2).

Feel free to change the PR target into the correct branch.

SeppahBaws commented:
Hmmm I'm also getting build errors when I check it out locally in a fresh clone

rgon commented:
The same happens with:
+ fresh clone
+ git checkout [6999e82](https://github.com/HoudiniGraphql/houdini/pull/1455/commits/6999e827941be2a02b34dc5c0600cb425b83f376)
+ `pnpm i && pnpm run build`.

Yet this exact run was successful before: https://github.com/HoudiniGraphql/houdini/actions/runs/14329494126

I don't have a clue

SeppahBaws commented:
I'm suspecting a weird dependency version of esbuild being introduced somewhere.
In the logs I can see the error coming from `esbuild@0.15.10`, but that's a version from 3 years ago.
I also have a bunch of other versions of it in my node_modules, so maybe they're not plaing nicely together:
![image](https://github.com/user-attachments/assets/5c85e8cf-8fa1-445e-b0e9-974ac032e801)

rgon commented:
@SeppahBaws how are you checking which version of the package logs that error?

Some googling led me to eslint errors as well. I have managed to update all versions (remaining in pnpm-lock: 0.24.2 & 0.25.2) by kamikaze-updating many dependencies in the monorepo. However, the same error still stands.
It may be related to esbuild configuration at `packages/_scripts/build.js#build`

SeppahBaws commented:
@rgon if you dig into the reported error in the CI, you can see the package version in the path of the code file that emits the error
![image](https://github.com/user-attachments/assets/dcfbe0c3-e661-47e8-9ccd-da87d43fecbf)

rgon commented:
Thanks!

Unfortunately happens all the way up to 0.25.2.
```
houdini-react:compile: ✘ [ERROR] Expected ">" but found "client"
houdini-react:compile: 
houdini-react:compile:     src/runtime/index.tsx:32:3:
houdini-react:compile:       32 │       client={client()}
houdini-react:compile:          │       ~~~~~~
houdini-react:compile:          ╵       >
houdini-react:compile: 
houdini-react:compile: ✘ [ERROR] Expected ">" but found "value"
houdini-react:compile: 
houdini-react:compile:     src/runtime/routing/Router.tsx:142:28:
houdini-react:compile:       142 │     <VariableContext.Provider value={variables}>
houdini-react:compile:           │                               ~~~~~
houdini-react:compile:           ╵                               >
houdini-react:compile: 
houdini-react:compile: ✘ [ERROR] Expected ">" but found "client"
houdini-react:compile: 
houdini-react:compile:     src/runtime/index.tsx:32:3:
houdini-react:compile:       32 │       client={client()}
houdini-react:compile:          │       ~~~~~~
houdini-react:compile:          ╵       >
houdini-react:compile: 
houdini-react:compile: ✘ [ERROR] Expected ">" but found "value"
houdini-react:compile: 
houdini-react:compile:     src/runtime/routing/Router.tsx:142:28:
houdini-react:compile:       142 │     <VariableContext.Provider value={variables}>
houdini-react:compile:           │                               ~~~~~
houdini-react:compile:           ╵                               >
houdini-react:compile: 
houdini-react:compile: Error: Build failed with 2 errors:
houdini-react:compile: src/runtime/index.tsx:32:3: ERROR: Expected ">" but found "client"
houdini-react:compile: src/runtime/routing/Router.tsx:142:28: ERROR: Expected ">" but found "value"
houdini-react:compile:     at failureErrorWithLog (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:1477:15)
houdini-react:compile:     at /home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:946:25
houdini-react:compile:     at runOnEndCallbacks (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:1317:45)
houdini-react:compile:     at buildResponseToResult (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:944:7)
houdini-react:compile:     at /home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:971:16
houdini-react:compile:     at responseCallbacks.<computed> (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:623:9)
houdini-react:compile:     at handleIncomingPacket (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:678:12)
houdini-react:compile:     at Socket.readFromStdout (/home/projects/houdini/node_modules/.pnpm/esbuild@0.25.2/node_modules/esbuild/lib/main.js:601:7)
houdini-react:compile:     at Socket.emit (node:events:518:28)
houdini-react:compile:     at addChunk (node:internal/streams/readable:561:12) {
houdini-react:compile:   errors: [Getter/Setter],
houdini-react:compile:   warnings: [Getter/Setter]
houdini-react:compile: }
houdini-react:compile:  ELIFECYCLE  Command failed with exit code 1.
houdini-react:compile: ERROR: command finished with error: command (/home/projects/houdini/packages/houdini-react) pnpm run compile exited (1)
```

SeppahBaws commented:
Hmmm strange...
I'm at work currently so I don't quite have a lot of time to look into this now, but I can take a proper look at it when I get home tonight

rgon commented:
Thanks @SeppahBaws. I have uploaded the test with with updated dependencies here https://github.com/rgon/houdini/tree/wip/vite6-hmr-debug
Note that no migration has been done, just version upgrades to debug the build.

SeppahBaws commented:
Ok I also cannot figure out what's going wrong here. According to the esbuild docs it should just work, but clearly it isn't right here. Maybe there's a config option somewhere that I'm missing? For the large part it seems like the config is just automagically correct.

AlecAivazis commented:
Ive run into this a few times lately when people have tried to set up their local dev environment. i'm trying to get to the bottom of it, ill report back if/when I find something

AlecAivazis commented:
So i was able to get builds passing again by pulling in `main`s pnpm lock file. i'm still unsure why there seems to be some incompatability with the latest esbuild but at least its passing now 🤷

rgon commented:
@AlecAivazis this is going to seem unbelievable, but I have `git pulled` and cannot reproduce the build, the same error appears.

Vite 6 imports quite a newer esbuild version, so forcing an unsynced pnpm-lock file could cause quite some havoc

rgon commented:
There has to be an incompatible auto-updating version specifier somewhere in `_scripts/package.json`, since I don't believe a dependency from other package would cause issues.

| dependency                | OK Version | NOK version |
|---------------------------|------------|-------------|
| commander                 | 9.4.1      | 9.4.0       |
| esbuild                   | 0.15.18    | 0.15.10     |
| esbuild-plugin-replace    | 1.2.0      | 1.3.0       |
| rollup                    | 4.28.1     | 4.30.1      |
| rollup-plugin-typescript2 | 0.34.1     | 0.34.0      |
| typescript                | 4.9.3      | 4.9.4       |

Resolved `_scripts` dependency versions from WORKING `pnpm-lock.yaml` from `main`:
```yaml
  packages/_scripts:
    dependencies:
      commander:
        specifier: ^9.4.0
        version: 9.4.1
      esbuild:
        specifier: ^0.15.10
        version: 0.15.18
      esbuild-plugin-alias:
        specifier: ^0.2.1
        version: 0.2.1
      esbuild-plugin-replace:
        specifier: ^1.2.0
        version: 1.3.0
      glob:
        specifier: ^11.0.1
        version: 11.0.1
      rollup:
        specifier: ^4.28.1
        version: 4.30.1
      rollup-plugin-typescript2:
        specifier: ^0.34.0
        version: 0.34.1(rollup@4.30.1)(typescript@4.9.4)
      typescript:
        specifier: ^4.9
        version: 4.9.4
```

Resolved `_scripts` dependency versions from NOT WORKNG  `pnpm-lock.yaml` from this branch:
```yaml
  packages/_scripts:
    dependencies:
      commander:
        specifier: ^9.4.0
        version: 9.4.0
      esbuild:
        specifier: ^0.15.10
        version: 0.15.10
      esbuild-plugin-alias:
        specifier: ^0.2.1
        version: 0.2.1
      esbuild-plugin-replace:
        specifier: ^1.2.0
        version: 1.2.0
      glob:
        specifier: ^11.0.1
        version: 11.0.1
      rollup:
        specifier: ^4.28.1
        version: 4.28.1
      rollup-plugin-typescript2:
        specifier: ^0.34.0
        version: 0.34.0(rollup@4.28.1)(typescript@4.9.3)
      typescript:
        specifier: ^4.9
        version: 4.9.3
```

rgon commented:
Well that's now fixed! (`pnpm i` working too).

What a sneaky issue. Out of those, an incompatible `rollup` version was to blame: versions between `>=4.28.1` and somewhere `<4.30.1` are problematic, though I can't see anything relevant in rollup's changelog. I have updated every instance to prevent further problems.

rgon commented:
The previous issue is now fixed, ready for review!

There is still a problem with `pnpm --filter e2e-react tests`. In my local machine, the error differs from what appears in CI:
<details><pre>
$ pnpm --filter e2e-react tests


> e2e-react@0.0.0 tests /home/projects/houdini/e2e/react
> pnpm run build && playwright test


> e2e-react@0.0.0 build /home/projects/houdini/e2e/react
> vite build

vite v6.2.4 building for production...
🎩 Generating runtime...
❌ Unexpected type: DateTime
✓ 0 modules transformed.
🎩 Generating Server Assets...
vite v6.2.4 building SSR bundle for production...
✓ 162 modules transformed.
✗ Build failed in 514ms
✗ Build failed in 873ms
error during build:
[houdini] [houdini] Expected ID to be a GraphQL nullable type.
file: /home/projects/houdini/e2e/react/src/routes/optimistic-keys/+page.tsx
file: /home/projects/houdini/e2e/react/src/routes/optimistic-keys/+page.tsx
[houdini] Expected ID to be a GraphQL nullable type.
file: /home/projects/houdini/e2e/react/src/routes/optimistic-keys/+page.tsx
    at devAssert (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/jsutils/devAssert.js:12:11)
    at new GraphQLNonNull (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/type/definition.js:395:32)
    at typeFromAST (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/utilities/typeFromAST.js:21:27)
    at TypeInfo.enter (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/utilities/TypeInfo.js:186:56)
    at Object.enter (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/utilities/TypeInfo.js:387:16)
    at Module.visit (/home/projects/houdini/node_modules/.pnpm/graphql@16.9.0/node_modules/graphql/language/visitor.js:197:21)
    at Config.localDocumentData (file:///home/projects/houdini/packages/houdini/build/vite-esm/index.js:76256:14)
    at Object.tag (file:///home/projects/houdini/packages/houdini-react/build/plugin-esm/index.js:89415:58)
    at Object.enter (file:///home/projects/houdini/packages/houdini-react/build/plugin-esm/index.js:88536:22)
    at AsyncWalker.visit (file:///home/projects/houdini/packages/houdini-react/build/plugin-esm/index.js:88377:26)
 ELIFECYCLE  Command failed with exit code 1.
/home/projects/houdini/e2e/react:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  e2e-react@0.0.0 tests: `pnpm run build && playwright test`
</pre></details>

Referring to 
https://github.com/rgon/houdini/blob/d67c11eb4654e8d1e6f3bd237a611e4e8f3e774a/e2e/react/src/routes/optimistic-keys/%2Bpage.tsx#L25-L39

@SeppahBaws you co-authored the graphql v16 update, I'd appreciate some help with this if possible! It's a bit outside of my knowledge

AlecAivazis commented:
Found some time today to get all of the versions in line and I fixed the api implementation. Everything is checking out now so I think this is good for a final review. Thanks again for all of the help @rgon

AlecAivazis commented:
Oh damn i just realized that this PR also effectively brought the 2.0 branch up to speed with the recent work on main. no wonder it was such a big effort. Double thanks @rgon ❤️

AlecAivazis reviewed (changes requested):
Overall this looks really good! Just a few small tweaks and comments

rgon commented on packages/houdini/src/vite/houdini.ts:
```diff
@@ -16,20 +22,102 @@ import {
+
+			// TODO: return tainted files from generate()
+			// Instead, walk over the entire houdini directory and invalidate all modules
+			const taintedModules: EnvironmentModuleNode[] = []
```
Yeah, I was wondering too, listing all files is not elegant.

> if we've computed this value, it will only change when new documents are discovered

The previous approach did not actually react on file changes, just on OS file change events which get triggered on write. And since the pipeline writes every file every time, independent of what has changed, this means effectively a full reload of all modules (this can be checked on the browser console, but vite 5 logging doesn't print it).

I suppose we could check what has changed: keep a dict of [filename] in $houdini and their hashes, and diff the list -> find environmentModuleNode by path (as I'm doing here).

However, I'm not sure it's worth it: the slow part is the actual code generation, so I feel like my efforts are best spent on v2! I've left this to #1456

The good thing, despite that, is that all built modules are replaced in one single event, so the actual toll on the CPU (and browser especially) is much less than the previous approach (it's just 1 swap). It does eat up comparatively more bandwidth, though (not a worry since 9/10 use cases are local, and houdini bundles are small anyways).

In order to only swap the files that had changed, the previous `vite-watch-and-run` caused 1 (actually 2) event per change, and this played dirty with Vite 6 for some reason, always triggering a full page reload but only on larger projects. Even though that trigger was the 50 parallel events, I never understood the real cause behind it. It's now fixed tho, so not to worry.

AlecAivazis replied:
yeaa i see what you mean. its probably not worth the headache just yet. I'm happy to revisit if we think its necessary once we get the incremental story working. And I also agree its not totally worth worrying about diffing the generated files, we can figure out the correct story once we integrate the new pipeline.

rgon commented on packages/houdini/src/vite/houdini.ts:
```diff
@@ -16,20 +22,102 @@ import {
+			}
+			lastHotUpdateEvent = {
+				environment: environment.name,
+				file,
```
No, I forgot them in from a previous test. Removed, but kept it as an object (for readability and future pipeline implementation).

rgon commented:
@AlecAivazis No worries! Yeah, figured out I'd get the branch up to speed. Houdini is excellent DevX, I hope it is of help!

johnwook commented:
I'm a big fan of Houdini (and dealing with HMR issues myself 🥲). Watching the discussions and the progress of this PR has been really inspiring. Thanks for all your hard work, and I'd love to contribute someday!

AlecAivazis commented:
While doing some final testing, I ran into something. If I introduce an error in a query then the hmr server seems to break. To reproduce this, i added an invalid field to a query. For example, i added `birthday` in the query defined in 
 `src/routes/blocking/query_with_blocking_disable/+page.gql`.

rgon commented:
Interesting. So `vite-watch-and-run` hid the errors, but vite crashes if thrown.  That was an easy fix:
```
[web] 🎩 🔄 bundle HMR rebuild...
[web] ❌ Encountered error in src/routes/blocking/query_with_blocking_disable/+page.gql
[web] Cannot query field "BRITHODAY" on type "User". Did you mean "birthDate"?
[web] 🎩 🔄 bundle HMR rebuild...
```

AlecAivazis reviewed (approved):
Awesome, seems like everything is in order. Thanks again!
